### ISSUE TITLE
The CI uses up a lot of resources

### ISSUE THREAD

**Issue by TheKnarf** [TheBevyFlock/bevy_new_2d#212]
![Screenshot 2024-08-01 at 14 41 38](https://github.com/user-attachments/assets/e867ab2a-9240-4409-bdb2-019bf9220da4)

A single game jam was all it took.

Too late I realized that I should disable the windows/linux/mac build since I only really needed the web build for the game jam. I think the template probably should disable those by default and require people to enable them if they need them.

Maybe there are other things one could do to reduce the amount of minutes it takes up? Maybe running `cargo fmt` and `cargo test` isn't necessary by default on main. Maybe one could run `cargo test` without triggering it parsing the docs folder? Are there other improvements one could make?

**Comment by TheKnarf** [comment]
![Screenshot 2024-08-01 at 14 47 00](https://github.com/user-attachments/assets/e2470ed6-0b80-4ea0-a6db-b96c0b2a91fc)

CI builds took almost 30 minutes to run.

![Screenshot 2024-08-01 at 14 47 39](https://github.com/user-attachments/assets/93f2afbf-2d53-4f49-a007-0dafbaca2e08)

Release build took over 30 minutes when building for all targets, less than 15 minutes after disabling everything except for the web build.

**Comment by janhohenheim** [comment]
We intentionally made it do as much stuff as possible, assuming that most users would open source their code by default (which means infinite GitHub actions) or be ready to deal either the consequences of having a private repository, i.e. fiddling with the workflow.

That said, it would be fairly easy to enable or disable the builds based on a `cargo generate` flag. How would you feel about that?

The builds should take like a minute at most when cached. Did something go wrong there?

**Comment by TimJentzsch** [comment]
Under the hood it uses the GitHub Actions cache, I'm not aware of any restrictions for private repositories.

**Comment by benfrankel** [comment]
Yeah I'm curious if caching wasn't working, or otherwise why it was taking so long. Looking at my own game's CI, I see this error when trying to save the cache (https://github.com/benfrankel/blobo_party/actions/runs/10139032444):

```
zstd: error 70 : Write error : cannot write block : No space left on device 
/usr/bin/tar: cache.tzst: Cannot write: Broken pipe
/usr/bin/tar: Child returned status 70
/usr/bin/tar: Error is not recoverable: exiting now
Warning: Failed to save: "/usr/bin/tar" failed with error: The process '/usr/bin/tar' failed with exit code 2
```

This started happening on the same CI run that spontaneously started recompiling from `log`, making CI take 5x longer. Which is very strange because the `Cargo.toml` / `Cargo.lock` weren't touched as part of that commit.

**Comment by mockersf** [comment]
GitHub runners for private repos are smaller than for public repos, except macOS hosts
private: https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for--private-repositories
public: https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories

**Comment by janhohenheim** [comment]
@mockersf thanks! That explains the performance issue then. Nothing we can fix, but it might be good to point it out in the docs.

**Comment by TimJentzsch** [comment]
Hmm thats interesting.
GitHub should automatically delete old caches when the cache limit is being exceeded, but maybe it has a delay.

Maybe we can start optimizing the size of the cache though.

**Comment by benfrankel** [comment]
So I noticed that in the CI run I linked, the `Docs` and `Tests` jobs actually restored their cache from the _Release workflow_ as a fallback. It looks like cache thrashing to me.

**Comment by benfrankel** [comment]
Well I think the fallback itself is okay and shouldn't increase cache size, the problem is that `Tests` was unable to save its own cache afterwards for some reason.

Unless using a fallback means that when the cache is saved at the end of the job, it will also include all the irrelevant stuff from the downloaded fallback?

**Comment by TimJentzsch** [comment]
> Well I think the fallback itself is okay and shouldn't increase cache size, the problem is that `Tests` was unable to save its own cache afterwards for some reason.
> 
> Unless using a fallback means that when the cache is saved at the end of the job, it will also include any irrelevant stuff it downloaded from the fallback, thus increasing cache sizes every time a fallback occurs? If that's the case... maybe it would be better to have no fallback at all.

Yes, the stuff from the fallbacks will also be saved again because its in the same folder.
But without fallbacks I think we will have a cache miss too often...
If we use the fallback from the same job it should be fine though.
Unrelated jobs can double the cache size in the worst case

**Comment by benfrankel** [comment]
Could there be a situation where cache Z is a fallback from cache Y, which is a fallback from cache X, ... causing a ballooning cache size over time? I wonder if there's a way to prune unused stuff from `target/`.

**Comment by TimJentzsch** [comment]
Yes, I think in theory that's possible.
We should adjust the fallback in `cargo-cache` a bit.

Otherwise I guess it's a good idea to clear your caches in CI occasionally.

**Comment by janhohenheim** [comment]
Is it alright if I move this issue to `cargo cache`?

**Comment by benfrankel** [comment]
IMO that makes sense but this issue should remain here as well, since we should consider the fact that private runners have worse performance + are not free.

**Comment by janhohenheim** [comment]
Would a solution be to ask you in `cargo generate` which builds you want?

**Comment by benfrankel** [comment]
Yes, and a note in the workflows doc that mentions the issue with private runners, so users hopefully see that when they're setting up CI.

**Comment by TimJentzsch** [comment]
I created an issue for this on the `cargo-cache` side: <https://github.com/Leafwing-Studios/cargo-cache/issues/22>

**Comment by benfrankel** [comment]
Here's another issue on the `cargo-cache` side for the root cause / optimizing the cache size down even further: https://github.com/Leafwing-Studios/cargo-cache/issues/24

**Comment by benfrankel** [comment]
Remaining tasks:
- [ ] Add a warning about private repos (costs, quota, smaller / slower runners)
- [ ] Add a flag to opt out of caching in release workflow (to save cache space)
